src/apis/orderbookapi.py

```diff
@@ -0,0 +1,123 @@
+"""
+OrderbookAPI for fetching relevant data using the CoW Swap Orderbook API.
+"""
+# pylint: disable=logging-fstring-interpolation
+
+from typing import Any, Optional
+import json
+import requests
+from src.helper_functions import get_logger
+from src.models import Trade, OrderExecution
+from src.constants import (
+    header,
+    REQUEST_TIMEOUT,
+    SUCCESS_CODE,
+    FAIL_CODE,
+)
+
+PROD_BASE_URL = "https://api.cow.fi/mainnet/api/v1/"
+BARN_BASE_URL = "https://barn.api.cow.fi/mainnet/api/v1/"
+
+
+class OrderbookAPI:
+    """
+    Class for fetching data from a Web3 API.
+    """
+
+    def __init__(self):
+        self.logger = get_logger()
+
+    def get_solver_competition_data(self, tx_hash: str) -> Optional[dict[str, Any]]:
+        """
+        Get solver competition data from a transaction hash.
+        The returned dict follows the schema outlined here:
+        https://api.cow.fi/docs/#/default/get_api_v1_solver_competition_by_tx_hash__tx_hash_
+        """
+        prod_endpoint_url = f"{PROD_BASE_URL}solver_competition/by_tx_hash/{tx_hash}"
+        barn_endpoint_url = f"{BARN_BASE_URL}solver_competition/by_tx_hash/{tx_hash}"
+        try:
+            json_competition_data = requests.get(
+                prod_endpoint_url,
+                headers=header,
+                timeout=REQUEST_TIMEOUT,
+            )
+            if json_competition_data.status_code == SUCCESS_CODE:
+                solver_competition_data = json.loads(json_competition_data.text)
+            elif json_competition_data.status_code == FAIL_CODE:
+                barn_competition_data = requests.get(
+                    barn_endpoint_url, headers=header, timeout=REQUEST_TIMEOUT
+                )
+                if barn_competition_data.status_code == SUCCESS_CODE:
+                    solver_competition_data = json.loads(barn_competition_data.text)
+                else:
+                    return None
+        except requests.exceptions.ConnectionError as err:
```

Comment:
Since I am a bit ignorant, can other exceptions happen from such a request?

Reply:
There might be other errors. I can change it to `requests.RequestException` which should be the base type for all exceptions thrown.

Reply:
Yeah I guess I was wondering whether we want to handle everything here, within the class (which I believe is the right thing to do), or whether whoever calls this function should be aware that it might raise an exception. I would be in favor of catching everything here.

Reply:
For now I want to handle everything in at this level. I was trying to be cautious in the sense of not just catching everything but being a bit specific so we learn what error types there are and can add them later explicitly. I did not intend to let the caller catch errors.

In the future it might be beneficial to have our own exception types instead of a None as return value.
Then code would be a bit less clean maybe, but not by much, and it is more powerful. The code fragment
```python
value = api.get_value()
if value is None:
    whatever
```
would turn into
```python
try:
    value = api.get_value()
except EBBOException:
    whatever # can depend on the type of error now
```

Reply:
Made the return type consistent for now.